src/connections.rs

```diff
@@ -240,38 +191,70 @@ impl IncomingMessages {
                 warn!("Failed to read incoming message on bi-stream: {}", err);
                 None
             }
-            Some(Ok((send, recv))) => Self::read_bytes(recv, max_msg_size)
+            Some(Ok((send, mut recv))) => read_bytes(&mut recv, max_msg_size)
                 .await
-                .map(|bytes| (bytes, send)),
+                .ok()
+                .map(|bytes| (bytes, send, recv)),
         }
     }
+}
 
-    // Read the message's bytes which size is capped
-    async fn read_bytes(recv: quinn::RecvStream, max_msg_size: usize) -> Option<Bytes> {
-        match recv.read_to_end(max_msg_size).await {
-            Ok(wire_bytes) => {
-                trace!("Got new message with {} bytes.", wire_bytes.len());
-                match WireMsg::from_raw(wire_bytes) {
-                    Ok(WireMsg::UserMsg(msg_bytes)) => Some(Bytes::copy_from_slice(&msg_bytes)),
-                    Ok(WireMsg::EndpointEchoReq) | Ok(WireMsg::EndpointEchoResp(_)) => {
-                        // TODO: handle the echo request/response message
-                        warn!("UNIMPLEMENTED: echo message type not supported yet");
-                        None
-                    }
-                    Err(err) => {
-                        warn!("Failed to parse received message bytes: {}", err);
-                        None
-                    }
-                }
-            }
-            Err(err) => {
-                warn!("Failed reading message's bytes: {}", err);
-                None
-            }
+/// Read the message's bytes which size is capped
+pub async fn read_bytes(recv: &mut quinn::RecvStream, _max_msg_size: usize) -> Result<Bytes> {
+    let mut data_len: [u8; 1] = [0; 1];
+    recv.read_exact(&mut data_len).await?;
+    let mut data: Vec<u8> = vec![0; data_len[0] as usize];
+    recv.read_exact(&mut data).await?;
+    trace!("Got new message with {} bytes.", data.len());
+    match WireMsg::from_raw(data)? {
+        WireMsg::UserMsg(msg_bytes) => Ok(Bytes::copy_from_slice(&msg_bytes)),
+        WireMsg::EndpointEchoReq | WireMsg::EndpointEchoResp(_) => {
+            // TODO: handle the echo request/response message
+            unimplemented!("echo message type not supported yet");
         }
     }
 }
 
+/// Helper to send bytes to peer using the provided stream.
+pub async fn send_msg(
```

Comment:
Should this be `pub`?

Reply:
I made it public temporarily. In the latest commits I've made it private again.